src/app/server/JointFabricDatastorage.h

```diff
@@ -0,0 +1,198 @@
+    }
+    FabricIndex GetAdministratorFabricIndex() { return mAdministratorFabricIndex; }
+
+    CHIP_ERROR SetAnchorNodeId(NodeId anchorNodeId)
+    {
+        mAnchorNodeId = anchorNodeId;
+        return CHIP_NO_ERROR;
+    }
+    NodeId GetAnchorNodeId() { return mAnchorNodeId; }
+
+    CHIP_ERROR SetAnchorVendorId(VendorId anchorVendorId)
+    {
+        mAnchorVendorId = anchorVendorId;
+        return CHIP_NO_ERROR;
+    }
+    VendorId GetAnchorVendorId() { return mAnchorVendorId; }
+
+    CHIP_ERROR AddPendingNode(FabricIndex fabricId, NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR UpdateNode(NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR RemoveNode(NodeId nodeId);
+    CHIP_ERROR RefreshNode(NodeId nodeId);
+
+    CHIP_ERROR SetNode(NodeId nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum state);
+
+    CHIP_ERROR AddGroupKeySetEntry(uint16_t groupKeySetId);
+    CHIP_ERROR AddGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet);
+    bool IsGroupKeySetEntryPresent(uint16_t groupKeySetId);
+    CHIP_ERROR RemoveGroupKeySetEntry(uint16_t groupKeySetId);
+    CHIP_ERROR UpdateGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet);
+
+    CHIP_ERROR AddAdmin(Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type & adminId);
+    bool IsAdminEntryPresent(NodeId nodeId);
+    CHIP_ERROR UpdateAdmin(NodeId nodeId, CharSpan friendlyName, ByteSpan icac);
+    CHIP_ERROR RemoveAdmin(NodeId nodeId);
+
+    const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type> & GetGroupKeySetList()
+    {
+        return mGroupKeySetList;
+    }
+    const std::vector<GenericDatastoreNodeInformationEntry> & GetNodeInformationEntries() { return mNodeInformationEntries; }
+    const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type> &
+    GetAdminEntries()
+    {
+        return mAdminEntries;
+    }
+
+    /**
+     * Used to notify of changes in the node list and more TODO.
+     */
+    class Listener
+    {
+    public:
+        virtual ~Listener() = default;
+
+        /**
+         * Notifies of a change in the node list.
+         */
+        virtual void MarkNodeListChanged() = 0;
+
+    private:
+        Listener * mNext = nullptr;
+
+        friend class JointFabricDatastorage;
+    };
+
+    /**
+     * Add a listener to be notified of changes in the Joint Fabric Datastorage.
+     *
+     * @param [in] listener  The listener to add.
+     */
+    void AddListener(Listener & listener);
+
+    /**
+     * Remove a listener from being notified of changes in the Joint Fabric Datastorage
+     *
+     * @param [in] listener  The listener to remove.
+     */
+    void RemoveListener(Listener & listener);
+
+private:
+    static constexpr size_t kMaxNodes = 32;
```

Comment:
Looking into the code I see that the `kMaxNodes` represents the maximum number of regular Nodes, the maximum number of Administrator, the maximum number of GroupKeySets. For sure the maximum values for these attributes should be different but I don't know what the max value should be. @chrisdecenzo do we need to specify this in the JF Datastore text?

Reply:
I think it needs to be a lot larger than 32. If you need to go with something, I would choose 128 or 256

Reply:
Thanks for the feedback. I've changed it to 256 in the latest commit.

Reply:
Thx for changing it to 256:

- As I said above, I think it makes sense to keep kMaxNodes for the maximum number of regular Nodes - 256
- Add a separate kMaxAdminNodes for the maximum number of Administrators (probably keeping it to 16/32 is acceptable)
- Add a separate kMaxGroupKeySet: in my opinion this should be dependant on a different define (e.g.: kMaxGroups) that represents the maximum number of groups. Then KMaxGroupKeySet should be a multiple of kMaxGroups. And probably kMaxGroups should be dependant on kMaxNodes.

Reply:
Added the following members to the JointFabricDatastore class.
- kMaxNodes = 256;
- kMaxAdminNodes  = 32;
- kMaxGroups = kMaxNodes / 16;
- kMaxGroupKeySet = kMaxGroups * 16;